### ISSUE TITLE
Some breakpoints not pausing debugger

### ISSUE THREAD

**Issue by bergwerf** [Dart-Code/Dart-Code#1443]
I set a breakpoint on a print function, the print is shown in the debugger output various times but the debugger never pauses. Logs: https://pastebin.com/3rQDJsqu
When I put a breakpoint on a much earlier point in my code it does pause so perhaps there is some change in the Dart VM related to JIT optimization, but thats just wild speculation. I can remember it working fine so far, my current dart version is `2.2.0-dev.1.1` which is already more than a month old.

**Comment by DanTup** [comment]
I wonder if it's because `rootLib` looks like this:

> "uri":"file:\/\/\/home\/herman\/repos\/qedb\/cfa\/bin\/cfa.dart"

But our breakpoint looks like this:

> "scriptUri":"package:cfa/src/script/solver.dart"

In #957 we changed to only sending breakpoints as `package:` URIs when we could since all of the tests now passed that way (there were some complications with trying to set multiple breakpoints in the VM from only a single breakpoint in Code). This was expected to be save because of https://github.com/dart-lang/sdk/issues/33076

I wonder if it's caused by relative path imports from outside of the lib folder.. I'll do some testing and see if I can repro and add tests.

**Comment by DanTup** [comment]
Yep, this seems to be caused by relative path imports from outside lib. Since the entry point is outside `lib` it isn't considered a `package:` uri, and therefore the relative path is also not a `package:uri`.

A quick workaround would be to change your imports in your `bin/` scripts to be `package:cfa/xxx` imports.

@lrhn @peter-ahe-google before I start multiplying up breakpoints again, is there a better fir for this in Dart? For example, could/should imports into a packages `lib` folder also be recognised and mapped to `package:` uris (for the same reasons #33076 was done)?

**Comment by DanTup** [comment]
Ok cool, I'll hold off doing anything here for now to see what comes out of that. @bergwerf switching to `package:` URIs in your `bin` folder would be the best workaround for now. If it turns out this can't be fixed in Dart, then I'll look at bringing back the behaviour to send two breakpoints (or, picking between file/package based on whether the script is inside `lib`).

**Comment by bergwerf** [comment]
I am already including my library in `bin/` using `package:` rather than a relative import.

**Comment by DanTup** [comment]
Uh, ok - that's weird :( I can only repro with a relative path in the `bin` script.

Are you able to share the project, or a small sample that reproduces it in that way? I'm assuming it's all a single project, the breakpoint isn't in an external package?

**Comment by bergwerf** [comment]
Yes it is one project. I believe I am doing `package:` imports everywhere except for some small utility files that are in the lib source (break point is not in one of those btw).
I could send an archive with the project to your email (I'm not currently publishing it).

**Comment by DanTup** [comment]
Sure, if you want to mail to dantup @ google and include whether there's anything specific I need to do to run it, I'll take a look.

**Comment by DanTup** [comment]
Ok, I think it's related to part/part-of somehow. I was able to strip it down to this:

### pubspec.yaml
```yaml
name:          breakpoint_test
version:       0.0.1
```

### bin/main.dart
```dart
import 'package:breakpoint_test/all_scripts.dart';

void main() {
  func();
}
```

### lib/all_scripts.dart
```dart
library foo;

part 'src/func.dart';
```

### lib/src/func.dart
```dart
part of foo;

void func() {
  print("Breakpoints don't work here!");
}
```

If I move `func` directly into `all_scripts` then it works fine.

[breakpoint_test.zip](https://github.com/Dart-Code/Dart-Code/files/2829151/breakpoint_test.zip)

@peter-ahe-google @siva Is there an explanation for this behaviour, or is it a bug? VS Code is sending the breakpoint using the URI `package:breakpoint_test/src/func.dart` and in Observatory if I run `info breakpoints` I get:

> Future breakpoint 2 at package:breakpoint_test/src/func.dart:4

I don't know why it's a "Future breakpoint" since the code has been loaded. Anything more I can do to troubleshoot?

**Comment by DanTup** [comment]
I've split the relative-path issue out to https://github.com/Dart-Code/Dart-Code/issues/1446 since the problem here seems to be different.

**Comment by a-siva** [comment]
It is a future breakpoint because the actual breakpoint is set when the function is actually compiled (i.e JITed).

**Comment by DanTup** [comment]
@a-siva In this case, the code has been executed (it's printed to the screen). It only happens when the code is included in a `part`/`part-of`. It feels like a bug, though I'm not sure if it's because of the URI/paths or something else. What's the best way to debug further?

**Comment by peter-ahe-google** [comment]
My guess it that there could be a flaw in how the VM maps fileUris to importUris (this is a concept from Kernel, let me know if I need to elaborate).

We only store a fileUri on `func`. So you have to translate that back to the importUri, and it is not sufficient to look at the enclosing library's importUri.

I think it would be more reliable to translate importUris to fileUris (that is, convert package-URIs to file-URIs), because the debug information is based on those.

**Comment by DanTup** [comment]
@peter-ahe-google 

> So you have to

I presume this means in the VM and not in Dart Code? Should I open an issue in /sdk about this?

**Comment by peter-ahe-google** [comment]
Yes, I think this may be something that needs tweaking in the Dart VM. I've reached out to @a-siva to discuss this more.
